README.txt
app_weather v0.9.3
=========== ======

This is a Domotion app obtaining the weather from OpenWeatherMap or Yahoo weather. Weather information can be used in Domotion to control devices.

Install:
--------

Weather app can be installed using: install.sh
Start as: sudo ./install.sh <arguments>
  <no argument>: install app_weather
  -u/ -U       : uninstall app_weather
  -h/ -H       : this help file
  -l/ -l       : install decentralized (on another location, '/opt/Domo_app_weather', when running standalone not on the Domotion system)
  -e/ -E       : uninstall decentralized (on another location)
  -c/ -C       : Cleanup compiled files in install folder
  <folder>     : as second argument, alternative installation folder (service needs to be adapted manually if used)

Weather app can be installed in the '/opt/Domotion/apps' folder and is totally handled by Domotion. If you require the app to be installed on a different system, it is installed in '/opt/Domo_app_weather' and is started by the service as a stand-alone app. The app connects to Domotion which runs a BDA (Basic Device access) server.

Domotion 1.2.0 or higher must be installed to be able to install apps and use then with Domotion.

Settings:
---------

Settings are stored in app_weather.xml, by default located at '/etc/Domotion', or alternatively (if no access to '/etc/Domotion') at the home folder.

The following settings can be altered:
<domoapp>
	<app>												--> App specific settings
		<location>New York, US</location> 				--> Location for the weather app to check, try at 'https://www.openweathermap.org' or 'https://www.yahoo.com/news/weather' if it works
		<unitsmetric>True</unitsmetric>					--> If True, then Celcius and metric units, else Farenheit and imperai units.
		<apikey></apikey>								--> apikey for openweathermap, obtain one at: 'https://home.openweathermap.org/users/sign_up' for a free api key
		<weatheryahoo>False</weatheryahoo>				--> If True, the yahoo is used, else openweathermap is used
		<conditionsdaily>								--> Daily conditions to set the shade parameter (parameter is set high if complies)
			<high>24</high>								--> Maximum temperature higher then .. degrees
			<low>-15</low>								--> Minimum temperature lower then .. degrees
		</conditionsdaily>
		<conditionscurrent>								--> Current conditions to set the shade parameter (parameter is set high if complies and not overruled by dailyconditions)
			<temp>24</temp>								--> Current temperature higher then .. degrees
			<condition1>								--> Subcondition, implements and function
				<precipation>1</precipation>			--> Precipation more than 1 ...
				<clouds>70</clouds>						--> Clouds more that 70 %
			</condition1>
		</conditionscurrent>
		<conditionsdailyfull>							--> Same as conditionsdaily, but to set the fullshade parameter
			<high>28</high>
			<low>-20</low>
		</conditionsdailyfull>
		<conditionscurrentfull>							--> Same as conditionscurrent, but to set the fullshade parameter
			<temp>28</temp>
		</conditionscurrentfull>
	</app>
	<common>											--> Common settings for apps
		<verbose>False</verbose>						--> Log verbosity, set to false except for debugging
		<server></server>								--> Domotion IP or name if not on local system
		<port>60004</port>								--> Port on which the BDA server (Domotion) is running
		<url>app_weather</url>							--> URL for BDA server (can be changed if more instances of the app are running)
		<interval>600</interval>						--> Interval for querying current weather
		<times>											--> Time(s) for querying 5 days forecast
			<time1>5:55</time1>							--> Time for querying 5 days forecast
		</times>
		<username></username>							--> Username, or leave empty if running on local system
		<password></password>							--> Password, or leave empty if running on local system
	</common>
</domoapp>

Logging:
--------

Weather app logs to app_weather.log, by default located at '/var/log', or alternatively (if no access to '/var/log') at the home folder.

Controls for Domotion:
----------------------

The following controls are sent to Domotion. They can be captured if required. Add a control/ sensor:
Type= Input from URL
DeviceURL= <myip>/app_weather (if url if xml file is set to a different name, enter this name here)
KeyTAG= current_temp (or whatever parameter you are interested in)
SensorType= Temperature (or anything else)
Poll= False

The following parameters are available:
current_temp
current_precipation
current_clouds
current_windspeed
current_windangle

n = 0..4
highn
lown
precipationn
cloudsn
windspeedn
windanglen

shade
fullshade

That's all for now ...

Please send Comments and Bugreports to hellyrulez@home.nl


CHANGELOG.txt
[B]0.9.3[/B]
- Changing installation for DPKG/ APT compatibility

[B]0.9.2[/B]
- Converted code to python 3

[B]0.9.1[/B]
- Added ConditionsxxxFull parameters to be able to shade fully or have 2 different shade thresholds

[B]0.9.0[/B]
- Initial release


install.sh
#!/bin/bash
INSTALL="/usr/bin/install -c"
INSTALL_DATA="$INSTALL -m 644"
INSTALL_PROGRAM="$INSTALL"
INSTALL_FOLDER="cp -r "
NAME="Domotion"
ETCDIR="/etc"
ETCLOC="$ETCDIR/$NAME"
OPTDIR="/opt"
OPTLOC="$OPTDIR/$NAME"
APPS="apps"
APPLOC="$OPTLOC/$APPS"
APPSNAME="DomoApps"
STARTAPP="StartApp"
SERV=".service"
SERVICEDIR="$ETCDIR/systemd/system"
APPSSERVICESCRIPT="$APPSNAME""$SERV"
PRE="app_"
PY=".py"
XML=".xml"
DEBFOLDER="debian"

if [ "$EUID" -ne 0 ]
then
	echo "Please execute as root ('sudo install.sh')"
	exit
fi

APPNAME=$(find ".$APPLOC" -name "$PRE"*"$PY")
APPBASE=$(basename "$APPNAME" "$PY")

if [ -z "$APPNAME" ]
then
	echo "No apps found, folder corrupted"
	exit
else
	echo "App found: $APPBASE"
fi

APPPY="$APPBASE""$PY"
APPXML="$APPBASE""$XML"
APPDIR="$APPBASE""d"
OPTLOC2="$OPTDIR""/Domo_""$APPBASE"
SERVICESCRIPT="Domo_""$APPBASE""$SERV"

if [ "$1" == "-u" ] || [ "$1" == "-U" ]
then
	echo "$APPBASE uninstall script"

	if [ "$#" -gt 1 ]; then
		OPTLOC="$2"
	fi

    APPSLOC="$OPTLOC/$APPS"

	if [ ! -d "$OPTLOC" ]; then
		echo "Domotion installation not found"
		echo "Try: ./install.sh -u <alternative folder name>"
		echo "See also: ./install.sh -h"
		exit
	fi

	if [ ! -d "$APPSLOC" ]; then
		echo "Domotion apps not found"
		echo "This Domotion version is probably not able to deal with apps"
		echo "or the selected installation folder is incorrect"
		exit
	fi

	echo "Uninstalling $APPBASE"
	systemctl stop $APPSSERVICESCRIPT

	echo "Uninstalling $APPXML"
	if [ -e "$APPSLOC/$APPXML" ]; then rm -f "$APPSLOC/$APPXML"; fi
	echo "Uninstalling $APPPY"
	if [ -e "$APPSLOC/$APPPY" ]; then rm -f "$APPSLOC/$APPPY"; fi
	echo "Uninstalling $APPDIR"
	if [ -d "$APPSLOC/$APPDIR" ]; then rm -rf "$APPSLOC/$APPDIR"; fi

	echo "Uninstalling finished, try restarting $APPSNAME service"
	systemctl start $APPSSERVICESCRIPT
elif [ "$1" == "-e" ] || [ "$1" == "-E" ]
then
	echo "$APPBASE uninstall decentralized script"

	if [ "$#" -gt 1 ]; then
		OPTLOC2="$2"
	fi

	if [ ! -d "$OPTLOC2" ]; then
		echo "Domotion installation not found"
		echo "Try: ./install.sh -e <alternative folder name>"
		echo "See also: ./install.sh -h"
		exit
	fi

	echo "Uninstalling daemon $SERVICESCRIPT"
	systemctl stop "$SERVICESCRIPT"
	systemctl disable "$SERVICESCRIPT"
	if [ -e "$SERVICEDIR/$SERVICESCRIPT" ]; then rm -f "$SERVICEDIR/$SERVICESCRIPT"; fi

	echo "Uninstalling $APPBASE"

	if [ -d "$OPTLOC2" ]; then rm -rf "$OPTLOC2"; fi

	echo "Uninstalling finished"

elif [ "$1" == "-h" ] || [ "$1" == "-H" ]
then
	echo "Usage:"
	echo "  <no argument>: install $APPBASE"
	echo "  -u/ -U       : uninstall $APPBASE"
	echo "  -h/ -H       : this help file"
    echo "  -d/ -D       : build debian package"
    echo "  -s/ -S       : build debian standalone package"
	echo "  -l/ -L       : install decentralized (on another location)"
	echo "  -e/ -E       : uninstall decentralized (on another location)"
	echo "  -c/ -C       : Cleanup compiled files in install folder"
	echo "  <folder>     : as second argument, alternative installation folder"
elif [ "$1" == "-c" ] || [ "$1" == "-C" ]
then
	echo "$APPBASE Deleting compiled files in install folder"
	py3clean .
    rm -f ./*.deb
    rm -rf "$DEBFOLDER"
elif [ "$1" == "-d" ] || [ "$1" == "-D" ]
then
	echo "$APPBASE build debian package"
    rm -rf "$DEBFOLDER"
    cp -R debianapp "$DEBFOLDER"
	py3clean .
	fakeroot debian/rules clean binary
	mv ../*.deb .
elif [ "$1" == "-s" ] || [ "$1" == "-S" ]
then
	echo "$APPBASE build debian standalone package"
    rm -rf "$DEBFOLDER"
    cp -R debiansta "$DEBFOLDER"
	py3clean .
	fakeroot debian/rules clean binary
	mv ../*.deb .
elif [ "$1" == "-l" ] || [ "$1" == "-L" ]
then
	echo "$APPBASE install decentralized script"

	if [ "$#" -gt 1 ]; then
		OPTLOC2="$2"
	fi

	if [ ! -d "$OPTLOC2" ]; then
		mkdir "$OPTLOC2"
		chmod 755 "$OPTLOC2"
	fi

	echo "Installing $APPBASE"

	py3clean .

	if [ -e "$APPSLOC/$APPXML" ]; then rm -f "$APPSLOC/$APPXML"; fi
	echo "Uninstalling $APPPY"
	if [ -e "$APPSLOC/$APPPY" ]; then rm -f "$APPSLOC/$APPPY"; fi
	echo "Uninstalling $APPDIR"
	if [ -d "$APPSLOC/$APPDIR" ]; then rm -rf "$APPSLOC/$APPDIR"; fi
	echo "Installing $APPXML"

	if [ ! -e "./$APPXML" ]; then
		echo "$APPXML not found for this app, skipping"
	else
		if [ ! -d "$ETCLOC" ]; then
			mkdir "$ETCLOC"
			chmod 755 "$ETCLOC"
		fi
		if [ ! -e "$ETCLOC/$APPXML" ]; then
			$INSTALL_DATA ".$ETCLOC/$APPXML" "$ETCLOC/$APPXML"
		fi
	fi

	echo "Installing $APPPY"

	if [ ! -e "./$APPPY" ]; then
		echo "$APPPY not found for this app, skipping"
	else
		$INSTALL_PROGRAM ".$APPLOC/$APPPY" "$OPTLOC2/$APPPY"
	fi

	echo "Installing $APPDIR"

	if [ ! -e "./$APPDIR" ]; then
		echo "$APPDIR not found for this app, skipping"
	else
		$INSTALL_FOLDER ".$APPLOC/$APPDIR" "$OPTLOC2/$APPDIR"
	fi

	echo "Installing $STARTAPP"

	if [ ! -e "./$STARTAPP" ]; then
		echo "$STARTAPP not found for this app, skipping"
	else
		$INSTALL_PROGRAM ".$OPT/$STARTAPP" "$OPTLOC2/$STARTAPP"
	fi

	echo "Installing text files"
	if [ -e "./$CHANGELOG.txt" ]; then
		$INSTALL_PROGRAM "./$CHANGELOG.txt" "$OPTLOC2/$CHANGELOG.txt"
	fi
	if [ -e "./$LICENSE.txt" ]; then
		$INSTALL_PROGRAM "./$LICENSE.txt" "$OPTLOC2/$LICENSE.txt"
	fi
	if [ -e "./$README.txt" ]; then
		$INSTALL_PROGRAM "./$README.txt" "$OPTLOC2/$README.txt"
	fi

	echo "Installing decentralized common folder"
	COMMONLOC="$OPTLOC2""/appcommon"
	if [ ! -d "$COMMONLOC" ]; then
		mkdir "$COMMONLOC"
		chmod 755 "$COMMONLOC"
	fi

	CURLOC=$(pwd)
	cd $COMMONLOC
	wget https://raw.githubusercontent.com/Helly1206/Domotion/master/apps/appcommon/appcommon.py
	wget https://raw.githubusercontent.com/Helly1206/Domotion/master/apps/appcommon/bdaclient.py
	wget https://raw.githubusercontent.com/Helly1206/Domotion/master/apps/appcommon/bdauri.py
	cd $CURLOC

	echo "Installing daemon $APPBASE"
	read -p "Do you want to install an automatic startup service for $APPBASE (Y/n)? " -n 1 -r
	echo    # (optional) move to a new line
	if [[ $REPLY =~ ^[Nn]$ ]]
	then
		echo "Skipping install automatic startup service for $APPBASE"
	else
		echo "Install automatic startup service for $APPBASE"
		$INSTALL_DATA ".$SERVICEDIR/$SERVICESCRIPT" "$SERVICEDIR/$SERVICESCRIPT"

		systemctl enable $SERVICESCRIPT
		systemctl start $SERVICESCRIPT
	fi

	echo "Installing finished"

else
	echo "$APPBASE install script"

	if [ "$#" -gt 1 ]; then
		OPTLOC="$2"
	elif [ "$#" -eq 1 ]; then
		OPTLOC="$1"
	fi

	if [ ! -d "$OPTLOC" ]; then
		echo "Domotion installation not found"
		echo "Try: ./install.sh -l <alternative folder name>"
		echo "See also: ./install.sh -h"
		exit
	fi

	APPSLOC="$OPTLOC/$APPS"

	if [ ! -d "$APPSLOC" ]; then
		echo "Domotion apps not found"
		echo "This Domotion version is probably not able to deal with apps"
		echo "or the selected installation folder is incorrect"
		exit
	fi

	echo "Installing $APPBASE"

	py3clean .

	echo "Installing $APPXML"

	if [ ! -e "./$APPXML" ]; then
		echo "$APPXML not found for this app, skipping"
	else
		if [ ! -d "$ETCLOC" ]; then
			mkdir "$ETCLOC"
			chmod 755 "$ETCLOC"
		fi
		if [ ! -e "$ETCLOC/$APPXML" ]; then
			$INSTALL_DATA ".$ETCLOC/$APPXML" "$ETCLOC/$APPXML"
		fi
	fi

	echo "Installing $APPPY"

	if [ ! -e "./$APPPY" ]; then
		echo "$APPPY not found for this app, skipping"
	else
		$INSTALL_PROGRAM ".$APPLOC/$APPPY" "$APPSLOC/$APPPY"
	fi

	echo "Installing $APPDIR"

	if [ ! -e "./$APPDIR" ]; then
		echo "$APPDIR not found for this app, skipping"
	else
		$INSTALL_FOLDER ".$APPLOC/$APPDIR" "$APPSLOC/$APPDIR"
	fi

	echo "Installation finished, try restarting $APPSNAME service to include and start app"
	systemctl stop $APPSSERVICESCRIPT
	systemctl start $APPSSERVICESCRIPT
fi


app_weather.py
# -*- coding: utf-8 -*-
#########################################################
# SERVICE : app_weather.py                              #
#           Python app for getting weather information  #
#           from Yahoo or OpenWeatherMap                #
#           I. Helwegen 2018                            #
#########################################################

####################### IMPORTS #########################
from appcommon.appcommon import appcommon
from app_weatherd.YahooWeather import YahooWeather
from app_weatherd.OpenWeatherMap import OpenWeatherMap
from requests import ConnectionError

#########################################################
# Class : appcommon                                     #
# Main function: Do NOT Change name                     #
#########################################################
# DON'T Change this code                                #
#########################################################
class app(appcommon):
    def __init__(self):
        super(app, self).__init__(__file__)

    def __del__(self):
        super(app, self).__del__()

#########################################################
# Implement your code here                              #
#########################################################
    def init(self):
        self.weatherinfo = None
        self.dailyshade = 0
        self.currentshade = 0
    
        self.logger.info("Starting app: app_weather")
        self.appParams=self.ReadAppParams()
        self.logger.debug(self.appParams)
        
        try:
            if self.str2bool(self.appParams['weatheryahoo']):
                self.weatherinfo = YahooWeather(self.getBaseName(), self.str2bool(self.appParams['unitsmetric']), self.appParams['location'], self.appParams['apikey'])
            else:
                self.weatherinfo = OpenWeatherMap(self.getBaseName(), self.str2bool(self.appParams['unitsmetric']), self.appParams['location'], self.appParams['apikey'])
        except KeyError:
            self.logger.error("Error: XML file error")
        except ConnectionError:
            self.logger.error("Error: Obtaining location info, connection error")
        except Exception as e:
            self.logger.exception(e)
        

    def loop(self):
        current = False
        try:
            if (self.update()):
                if self.isIntervalUpdate():
                    current = True
                    params = self.weatherinfo.GetCurrent()
                    self.currentshade = self.CurrentShade(self.appParams, params)
                else:
                    params = self.weatherinfo.Get5dayForecast()
                    self.dailyshade = self.DailyShade(self.appParams, params)
                if params:
                    params['shade']=(self.dailyshade>0) or (self.currentshade>0)
                    params['fullshade']=(self.dailyshade>1) or (self.currentshade>1)
                    self.logger.debug(params)
                    self._loginfo(params, current)
                    self.setparameters(params)
        except KeyError:
            self.logger.error("Error: XML file error")
        except ConnectionError:
            self.logger.error("Error: Obtaining weather info, connection error; current:%d; retry in 1m", current)
            self.tryAgainIn1Minute(current)
        except Exception as e:
            self.logger.exception(e)

    def exit(self):
        del self.weatherinfo
        self.logger.info("Ready")

    def callback(self, tag, value):
        rtag = None
        rvalue = None
        if tag:
            if not value and params: # set is not allowed, get is
                if tag in list(params.keys()):
                    rtag = tag
                    rvalue = params[tag]
        return rtag, rvalue

#########################################################
# Local methods to implement                            # 
######################################################### 
    def _loginfo(self, params, current):
        if current:
            self.logger.info("T:%d, p:%f, c:%d, w:%d;%d s:%d;%d;%d", params['current_temp'], params['current_precipation'], params['current_clouds'], params['current_windspeed'], params['current_windangle'], self.currentshade, params['shade'], params['fullshade'])
        else:
            self.logger.info("Tl:%d, Th: %d, p:%f, c:%d, w:%d;%d s:%d;%d;%d", params['low0'], params['high0'], params['precipation0'], params['clouds0'], params['windspeed0'], params['windangle0'], self.dailyshade, params['shade'], params['fullshade'])

    def _getKey(self, params, key):
        retval = None
        for paramkey in list(params.keys()):
            if key in paramkey:
                retval = paramkey
    
        return retval
    
    def _getShadeParam(self, shadeParams, key, params):
        retval = False
    
        if key == "low":
            if self._getKey(params, key):
                try:
                    retval = (float(params[self._getKey(params, key)]) <= float(shadeParams[key]))
                except:
                    pass
        else:
            if self._getKey(params, key):
                try:
                    retval = (float(params[self._getKey(params, key)]) >= float(shadeParams[key]))
                except:
                    pass
    
        return retval
    
    def _getShade(self, shadeParams, params):
        shade = False
    
        for param in list(shadeParams.keys()):
            if isinstance(shadeParams[param],dict):
                _shade = True
                count = 0
                for subparam in list(shadeParams[param].keys()):
                    _shade = _shade and self._getShadeParam(shadeParams[param], subparam, params)
                    count += 1
                if count > 0:
                    shade = shade or _shade
            else:
                shade = shade or self._getShadeParam(shadeParams, param, params)
    
        return shade
    
    def DailyShade(self, appParams, params):
        shade = 0
        todaysparams = {}
    
        #get only today
        for key in list(params.keys()):
            if key[-1] == '0':
                todaysparams[key] = params[key]
    
        if 'conditionsdaily' in appParams:
            if self._getShade(appParams['conditionsdaily'], todaysparams):
                shade = 1
        if 'conditionsdailyfull' in appParams:
            if self._getShade(appParams['conditionsdailyfull'], todaysparams):
                shade = 2
    
        return shade
    
    def CurrentShade(self, appParams, params):
        shade = False
        if 'conditionscurrent' in appParams:
            if self._getShade(appParams['conditionscurrent'], params):
                shade = 1
        if 'conditionscurrentfull' in appParams:
            if self._getShade(appParams['conditionscurrentfull'], params):
                shade = 2
    
        return shade

#########################################################
# Main for local calling                                # 
#########################################################     
# Do NOT Change code to be able to perform autostarting #
#########################################################
if __name__ == "__main__":
    app().ismain()


opt/Domotion/apps/app_weatherd/YahooWeather.py
# -*- coding: utf-8 -*-
#########################################################
# SERVICE : YahooWeather.py                             #
#           Service that gives weather information      #
#           from the yahoo weather service              #
#########################################################

####################### IMPORTS #########################
from builtins import str
from builtins import range
from builtins import object
import requests, urllib.request, urllib.parse, urllib.error, json
import logging

#########################################################

####################### GLOBALS #########################
NO_LOC = -10000
#########################################################

###################### FUNCTIONS ########################

#########################################################

#########################################################
# Class : YahooWeather                                  #
#########################################################
class YahooWeather(object):
    baseurl = "https://query.yahooapis.com/v1/public/yql?"
    def __init__(self, basename, unitsMetric, LocationStr, ApiKey = None):
        self.verbose=True
        self.logger = logging.getLogger('{}.YahooWeather'.format(basename))
        if unitsMetric:
            self.units='c'
        else:
            self.units='f'
        self.ApiKey = ApiKey
        self.LocationStr = LocationStr
        self.woeid = NO_LOC
        try:
            self.GetLocation(LocationStr)
        except:
            pass

    def __del__(self):
        pass

    def GetLocation(self, LocationStr):
        woeid_query='select woeid from geo.places(1) where text="{}"'.format(LocationStr)
        yql_url = self.baseurl + urllib.parse.urlencode({'q':woeid_query}) + "&format=json"
        result = requests.get(yql_url)
        data = json.loads(result.text)

        try:
            if data['query']['count']>1:
                self.logger.error("Location cannot be found unambiguously, exiting ...")
                return False
            else:
                self.woeid=data['query']['results']['place']['woeid']
                self.logger.info("Obtained location data woeid = %s", self.woeid)
        except:
            return False

        return True

    def Get5dayForecast(self):
        params = {}
        doForecast = True
        if self.woeid == NO_LOC:
            doForecast = self.GetLocation(self.LocationStr)
        if doForecast:
            if not self.units:
                yql_unit=""
            else:
                yql_unit=" and u='{}'".format(self.units.lower())

            try:
                yql_query = "select item.forecast from weather.forecast where woeid={}{}".format(self.woeid,yql_unit)
                yql_url = self.baseurl + urllib.parse.urlencode({'q':yql_query}) + "&format=json"
                result = requests.get(yql_url)
                data = json.loads(result.text)
                forecast=data['query']['results']['channel']

                self.logger.info("Queried: 5 day forecast")
                l = len(forecast)
                if l>5:
                    l=5
                for i in range (0, l):
                    if self.verbose:
                        self.logger.debug("Date            : %s","{} {}".format(forecast[i]['item']['forecast']['day'],forecast[i]['item']['forecast']['date']))
                        self.logger.debug("High Temperature: %d",int(round(float(forecast[i]['item']['forecast']['high']))))
                        self.logger.debug("Low Temperature : %d",int(round(float(forecast[i]['item']['forecast']['low']))))
                        self.logger.debug("Precipation     : %f",self._getPrecipation(forecast[i]['item']['forecast']['text']))
                        self.logger.debug("Clouds          : %d",self._getClouds(forecast[i]['item']['forecast']['text']))
                        self.logger.debug("Wind speed      : %d",'0')
                        self.logger.debug("Wind angle      : %d",'0')
                        self.logger.debug("-------------------")
                    params['high'+str(i)] = int(round(float(forecast[i]['item']['forecast']['high'])))
                    params['low'+str(i)] = int(round(float(forecast[i]['item']['forecast']['low'])))
                    params['precipation'+str(i)] = self._getPrecipation(forecast[i]['item']['forecast']['text'])
                    params['clouds'+str(i)] = self._getClouds(forecast[i]['item']['forecast']['text'])
                    params['windspeed'+str(i)] = 0 # unavailable
                    params['windangle'+str(i)] = 0 # unavailable
            except:
                pass

        return params

    def GetCurrent(self):
        params = {}
        doForecast = True
        if self.woeid == NO_LOC:
            doForecast = self.GetLocation(self.LocationStr)
        if doForecast:
            if not self.units:
                yql_unit=""
            else:
                yql_unit=" and u='{}'".format(self.units.lower())

            try:
                yql_query = "select item.condition,wind from weather.forecast where woeid={}{}".format(self.woeid,yql_unit)
                yql_url = self.baseurl + urllib.parse.urlencode({'q':yql_query}) + "&format=json"
                result = requests.get(yql_url)
                data = json.loads(result.text)

                forecast=data['query']['results']['channel']

                self.logger.info("Queried: Current weather")
                if self.verbose:
                    self.logger.debug("Date               : %s","{}".format(forecast['item']['condition']['date']))
                    self.logger.debug("Current Temperature: %d",int(round(float(forecast['item']['condition']['temp']))))
                    self.logger.debug("Current Precipation: %f",self._getPrecipation(forecast['item']['condition']['text']))
                    self.logger.debug("Current Clouds     : %d",self._getClouds(forecast['item']['condition']['text']))
                    self.logger.debug("Current Wind speed : %d",int(round(float(forecast['wind']['direction']))))
                    self.logger.debug("Current Wind angle : %d",int(round(float(forecast['wind']['speed']))))
                    self.logger.debug("-------------------")
                params['current_temp'] = int(round(float(forecast['item']['condition']['temp'])))
                params['current_precipation'] = self._getPrecipation(forecast['item']['condition']['text'])
                params['current_clouds'] = self._getClouds(forecast['item']['condition']['text'])
                params['current_windspeed'] = int(round(float(forecast['wind']['speed'])))
                params['current_windangle'] = int(round(float(forecast['wind']['direction'])))
            except:
                pass

        return params

    def _getPrecipation(self, input):
        # precipation is unavialable, so lets guess from text information
        precip = 0
        p1 = ['drizzle', 'light snow showers', 'blowing snow', 'sleet', 'haze']
        p2 = ['mixed', 'freezing rain', 'snow flurries', 'hail', 'isolated thunderstorms', 'scattered', 'thundershowers', 'snow showers']
        p3 = ['showers', 'snow']
        p4 = ['tornado', 'tropical storm', 'hurricane', 'thunderstorms', 'heavy snow']
        p6 = ['severe thunderstorms']
        if any(x in input.lower() for x in p1):
            precip = 1
        elif any(x in input.lower() for x in p2):
            precip = 2
        elif any(x in input.lower() for x in p3):
            precip = 3
        elif any(x in input.lower() for x in p4):
            precip = 4
        elif any(x in input.lower() for x in p6):
            precip = 6

        return precip

    def _getClouds(self, input):
        # cloud information is unavialable, so lets guess from text information
        clouds = 100
        c20 = ['fair']
        c50 = ['tornado', 'tropical storm', 'hurricane', 'dust', 'smoky', 'blustery', 'windy', 'partly cloudy']
        c80 = ['mostly cloudy', 'scattered thunderstorms']
        c0 = ['not available', 'cold', 'clear', 'sunny', 'hot']
        if any(x in input.lower() for x in c20):
            clouds = 20
        elif any(x in input.lower() for x in c50):
            clouds = 50
        elif any(x in input.lower() for x in c80):
            clouds = 80
        elif any(x in input.lower() for x in c0):
            clouds = 0

        return clouds

"""                  precip[mm] clouds[%]
0   tornado                 4   50
1   tropical storm          4   50
2   hurricane               4   50
3   severe thunderstorms    6   100
4   thunderstorms           4   100
5   mixed rain and snow     2   100
6   mixed rain and select   2   100
7   mixed snow and select   2   100
8   freezing drizzle        1   100
9   drizzle                 1   100
10  freezing rain           2   100
11  showers                 3   100
12  showers                 3   100
13  snow flurries           2   100
14  light snow showers      1   100
15  blowing snow            1   100
16  snow                    3   100
17  hail                    2   100
18  sleet                   1   100
19  dust                    0   50
20  foggy                   0   100
21  haze                    1   100
22  smoky                   0   50
23  blustery                0   50 
24  windy                   0   50 
25  cold                    0   0
26  cloudy                  0   100
27  mostly cloudy (night)   0   80
28  mostly cloudy (day)     0   80
29  partly cloudy (night)   0   50
30  partly cloudy (day)     0   50
31  clear (night)           0   0
32  sunny                   0   0
33  fair (night)            0   20
34  fair (day)              0   20  
35  mixed rain and hail     2   100
36  hot                     0   0
37  isolated thunderstorms  2   100
38  scattered thunderstorms 2   80
39  scattered thunderstorms 2   80
40  scattered showers       2   100
41  heavy snow              4   100
42  scattered snow showers  2   100
43  heavy snow              4   100
44  partly cloudy           0   50
45  thundershowers          2   100
46  snow showers            2   100
47  isolated thundershowers 2   100
3200    not available       0   0
"""